ISPN-14592 Hot Rod client uses generated parser

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-14592

The parser can live with the already existing parser. Currently, it can handle the parsing of topology, exceptions, and ping.

jabolina commented:
Only the new API uses the generated client. There are a few changes on the `hotrod-client` to use some classes from the new package.

jabolina commented:
Currently, the parser can handle get, put, set, and ping. Additionally, handle exceptions and topology updates.

wburns commented on client/hotrod-client-transport/src/main/resources/hotrod.gr:
```diff
@@ -0,0 +1,245 @@
+keyType: mediaType;
+valueType: mediaType;
+mediaType returns MediaType
+   : { delegate.currentProtocolVersionIsOlder(2, 9) >= 0 }? mediaTypeDefinition mediaTypeDescription { mediaTypeDescription }
```
I believe the new parser only supports HR 4 right? Might as well rip out the old stuff at the same time.

wburns commented on client/hotrod-client-transport/src/main/resources/hotrod.gr:
```diff
@@ -0,0 +1,245 @@
+pVersion: uByte;
+negotiatedProtocolVersion: protocolVersion;
+protocolVersion returns short
+    : { delegate.protocolVersionIsOlder(3, 0) < 0 }? { delegate.getProtocolVersion() }
```
Same here, should only require HR 4 and newer afaik.

wburns commented on client/hotrod-client-transport/pom.xml:
Since we are only supporting the new parser in the new client, there is no reason to put this new module. Just integrate this module into the hotrod module directly, which means we don't have to move all these other classes as well or change the hotrod-client module at all.

wburns reviewed (changes requested):
Moving all the packages will change the review quite a bit. So I will wait to do a deeper dive after that is changed. It should be a lot easier to review as well then :)

jabolina commented on client/hotrod-client-transport/pom.xml:
```diff
@@ -0,0 +1,97 @@
+               <files>
+                  <file>${basedir}/src/main/resources/hotrod.gr</file>
+               </files>
+               <userSwitchThreshold>10</userSwitchThreshold>
```
The generated method was not returning properly. I think it was fixed with https://github.com/infinispan/infinispan-maven-plugins/pull/6.

jabolina commented on client/hotrod-client-transport/src/main/resources/hotrod.gr:
```diff
@@ -0,0 +1,245 @@
+    }
+}
+
+deadend {
```
Tested it now, and it is throwing an NPE. Definitely broken.

jabolina replied:
Updated it so we complete the operation exceptionally.

jabolina commented:
A big update, which reduced the changes by a lot. Everything now is executed directly on the `infinispan-hotrod`, no new module. Focusing only on Hot Rod 4.0+ so could reduce the file, too. And applying all the suggestions, fixing deadend execution, and renaming the classes.

wburns commented on client/hotrod/src/main/java/org/infinispan/hotrod/impl/transport/netty/ByteBufUtil.java:
```diff
@@ -23,9 +23,44 @@ public static byte[] readArray(ByteBuf buf) {
+         bf.resetReaderIndex();
+         return null;
+      } else {
+         if (length == 0) {
```
This if block here should be outside of the outer if I think. If the last values on the ByteBuf was a zero sized array it would never finish I believe.

wburns commented on client/hotrod/src/main/java/org/infinispan/hotrod/impl/transport/netty/ByteBufUtil.java:
```diff
@@ -23,9 +23,44 @@ public static byte[] readArray(ByteBuf buf) {
+   public static byte[] readMaybeArray(ByteBuf bf) {
+      int length = readMaybeVInt(bf);
+      if (length == Integer.MIN_VALUE) {
+         return null;
```
This should check the reader index moved instead. Same issue we had in the RESP parser you mentioned.

wburns commented on client/hotrod/src/main/java/org/infinispan/hotrod/impl/transport/netty/ByteBufUtil.java:
```diff
@@ -23,9 +23,44 @@ public static byte[] readArray(ByteBuf buf) {
+
+   public static String readString(ByteBuf bf) {
+      int length = readMaybeVInt(bf);
+      if (length == Integer.MIN_VALUE) {
```
This should check the read buffer offset as well.

jabolina commented on client/hotrod/src/main/java/org/infinispan/hotrod/impl/transport/netty/ByteBufUtil.java:
```diff
@@ -101,6 +136,25 @@ public static long readVLong(ByteBuf buf) {
    }
 
+   public static long readMaybeVLong(ByteBuf buf) {
+      if (buf.readableBytes() > 0) {
```
The markReader was invoked in the caller. I can move this to `Intrinsics`, since it is only used in one place.

jabolina commented:
Updated with all the fixes. The `readVLong` was moved to `Intrinsics`, as this was the only location used. For the other methods, on the client side, I've moved the `markReaderIndex` call to be in the method body instead of the caller's.

I've applied some of the fixes on the server side, too. If you prefer, I can create a JIRA and add it here.

wburns commented:
Different PR just to keep the client and server code separated. So I would revert the ExtendedByteBufJava changes

wburns reviewed (approved):
LGTM, just rip out the server pieces

wburns commented:
Just waiting on CI to complete.

wburns commented:
Integrated into main, thanks @jabolina !
